src/Membership.sol

```diff
@@ -0,0 +1,427 @@
+        return userMessageLimit >= minRateLimitPerMembership && userMessageLimit <= maxRateLimitPerMembership;
+    }
+
+    /// @dev acquire a membership and trasnfer the fees to the contract
+    /// @param _sender address of the owner of the new membership
+    /// @param _idCommitment the idcommitment of the new membership
+    /// @param _rateLimit the user message limit
+    /// @param _numberOfPeriods the number of periods the user wants to acquire
+    /// @return index the index in the merkle tree
+    /// @return reusedIndex indicates whether a new leaf is being used or if using an existing leaf in the merkle tree
+    function _acquireMembership(
+        address _sender,
+        uint256 _idCommitment,
+        uint32 _rateLimit,
+        uint32 _numberOfPeriods
+    )
+        internal
+        returns (uint32 index, bool reusedIndex)
+    {
+        if (_numberOfPeriods == 0) revert NumberOfPeriodsCantBeZero();
+
+        (address token, uint256 amount) = priceCalculator.calculate(_rateLimit, _numberOfPeriods);
+        (index, reusedIndex) =
+            _setupMembershipDetails(_sender, _idCommitment, _rateLimit, _numberOfPeriods, token, amount);
+        _transferFees(_sender, token, amount);
+    }
+
+    function _transferFees(address _from, address _token, uint256 _amount) internal {
+        if (_token == address(0)) {
+            if (msg.value != _amount) revert IncorrectAmount();
+        } else {
+            if (msg.value != 0) revert OnlyTokensAccepted();
+            IERC20(_token).safeTransferFrom(_from, address(this), _amount);
+        }
+    }
+
+    /// @dev Setup a new membership. If there are not enough remaining rate limit to acquire
+    /// a new membership, it will attempt to erase existing memberships and reuse one of the
+    /// slots helds by the membership
+    /// @param _sender holder of the membership. Generally `msg.sender`
+    /// @param _idCommitment IDCommitment
+    /// @param _rateLimit User message limit
+    /// @param _numberOfPeriods the number of periods the user wants to acquire
+    /// @param _token Address of the token used to acquire the membership
+    /// @param _amount Amount of the token used to acquire the membership
+    /// @return index membership index on the merkle tree
+    /// @return reusedIndex indicates whether the index returned was a reused slot on the tree or not
+    function _setupMembershipDetails(
+        address _sender,
+        uint256 _idCommitment,
+        uint32 _rateLimit,
+        uint32 _numberOfPeriods,
+        address _token,
+        uint256 _amount
+    )
+        internal
+        returns (uint32 index, bool reusedIndex)
+    {
+        if (_rateLimit < minRateLimitPerMembership || _rateLimit > maxRateLimitPerMembership) {
+            revert InvalidRateLimit();
+        }
+
+        // Attempt to free expired membership slots
+        while (totalRateLimitPerEpoch + _rateLimit > maxTotalRateLimitPerEpoch) {
```

Comment:
This will try to remove expired memberships sequentially, but it's possible that the accumulated rate limit of all the memberships being removed is not enough for the chosen rate limit to acquire. Should we also accept an optional list of memberships to expire, so the user of the contract can explore the contract offchain and pass a list expired memberships whose accumulated rate limits is enough to acquire the rate limit they need?

Reply:
Something interesting here is to prefer inserting new elements on the tree rather than attempt to erase expired memberships first, mainly because there's no guarantee that an expired membership rate limit will match the rate limit to acquire. (and then having to iterate  to obtain enough rate limit from expired memberships).

Reply:
> there's no guarantee that an expired membership rate limit will match the rate limit to acquire

Is it technically possible to track / predict how much expired rate limit capacity we will have at a given point in time? (So that we can quickly tell if it's even worth it to start iterating through expired membership in an attempt to overwrite them, or their rate limit won't suffice anyway.)

Reply:
A proposal we discussed in a call was that:
- optional "slot to override" parameter when doing the insertion
- if no _slot to override_ specified, then it tries to grab next available slot in tree. If global rate limit gets exceeded, it fials
- if _slot to override_ is specified:
    - if _slot to override_ is not over grace period, fails
    - if _slot to override rate limit_ < _rate limit of new membership_, fails
   Otherwise, replace slot.
   
 This means that in terms of gas cost, it's more predictable:
 
- there is space, low gas cost
- no space, most specific membership to override, higher gas cost

than the previous idea of "override if there is something to override, otherwise do not".

Also, we should have a call on contract to purge memberships over grace period. so we can be the one to cover the purge cost.

Reply:
> Is it technically possible to track / predict how much expired rate limit capacity we will have at a given point in time? (So that we can quickly tell if it's even worth it to start iterating through expired membership in an attempt to overwrite them, or their rate limit won't suffice anyway.)

I'd say this can be done offchain and then via function parameters indicate to the contract whether we want to: just insert a new leaf in the tree / have the contract remove automatically expired / remove specific memberships. Then nwaku could call `estimateGas` on any of these options and choose the one that incurs in less gas cost.

> A proposal we discussed in a call ...

Ah! this seems to be aligned with what I just mentioned

> TLDR: do we need to  grab next available slot in tree every time, or is it perhaps better to do this (more) lazily?

This is hard to answer without testing.  Perhaps we can have the three options I mentioned earlier, and have some script to try to determine the behavior of the contract once we have a large enough number of memberships.

Reply:
@fryorcraken, @s-tikhomirov , 
In https://github.com/waku-org/waku-rlnv2-contract/pull/13/commits/8d43448de59a27e860576073753274f1ca6ecf65 the `register` function was modified and now exists as two  overloaded functions:
```
register(uint256 idCommitment, uint32 userMessageLimit);
register(uint256 idCommitment, uint32 userMessageLimit, uint256[] calldata membershipsToExpire)
```
I'm still working on the test units for the last function but this should cover what was discussed before.

Reply:
> The idea of slot overwriting was initially meant to save contract storage space

Interesting, for me it was about booting off the network publishers with expired membership to keep network traffic and membership set closely related.

> every time, or is it perhaps better to do this (more) lazily?

As @richard-ramos ' stated:

```
register(uint256 idCommitment, uint32 userMessageLimit);
register(uint256 idCommitment, uint32 userMessageLimit, uint256[] calldata membershipsToErase)
```

With 2 functions on the contrat, it is now possible for the app to provide more feedback:

1. App sees there is space on the tree (`R_{max}` not reached yet, including new membership).
2. App can use `register(uint256 idCommitment, uint32 userMessageLimit);` and only an insertion happen

else

1. App sees there is **no** space on the tree (`R_{max}` is reached when including new membership).
2. App finds an `Expired` membership with greater or equal rate limit (a good actor would actually use the expired membership with greatest rate limit)
3. App can warn user that membership insertion would cost a bit more than usual
4. Insertion can be done with `register(uint256 idCommitment, uint32 userMessageLimit, uint256[] calldata membershipsToErase)`

This gives more predictability to the user.
App devs can also decide to spend the gas to purge `Expired` memberships to make it cheaper for users.